### PR TITLE
encryption of password added

### PR DESCRIPTION
**Pull request by SetMyDream**

### CODE REVIEW

File: src/main/java/com/internet/shop/util/HashUtil.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        StringBuilder hashedPassword = new StringBuilder();
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance("SHA512");
+            messageDigest.update(salt);
+            byte[] digest = messageDigest.digest(password.getBytes());
+            for (byte b : digest) {
+                hashedPassword.append(String.format("%2x", b));
+            }
+        } catch (NoSuchAlgorithmException e) {
+            logger.error(e);
```
Comment (TimoxaV) [inline src/main/java/com/internet/shop/util/HashUtil.java:29]: checklist says logging is not enough here. Programm must stop if there is an error in hashing

File: src/main/resources/init_dbscheme [hunk 2]
```diff
@@ -79,6 +79,7 @@ CREATE TABLE `internet_shop`.`users`
     `username` varchar(256) NOT NULL,
     `login`    varchar(256) NOT NULL,
     `password` varchar(256) NOT NULL,
+    `salt` BIT(16) NULL DEFAULT NULL,
```
Comment (TimoxaV) [inline src/main/resources/init_dbscheme]: every user must have salt, so it can't be null

File: pom.xml [hunk 3]
```diff
@@ -33,6 +33,11 @@
             <artifactId>mysql-connector-java</artifactId>
             <version>8.0.21</version>
         </dependency>
+        <dependency>
+            <groupId>log4j</groupId>
+            <artifactId>log4j</artifactId>
+            <version>1.2.17</version>
+        </dependency>
```
Comment (TimoxaV) [inline pom.xml:40]: why do you need this dependency?

File: src/main/java/com/internet/shop/model/User.java [hunk 4]
```diff
@@ -15,11 +16,20 @@ public User(String name, String login, String password) {
...
     }
 
-    public User(Long id, String name, String login, String password) {
+    public User(Long id, String name, String login, String password, byte[] salt) {
         this.userId = id;
         this.name = name;
         this.login = login;
         this.password = password;
+        this.salt = salt;
+    }
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/model/User.java:23]: image
let's modify these constructors like this

File: src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java [hunk 5]
```diff
@@ -15,7 +16,8 @@
     public User login(String login, String password) throws AuthenticationException {
         User user = userService.findByLogin(login).orElseThrow(() ->
                 new AuthenticationException("Wrong login or password"));
-        if (user.getPassword().equals(password)) {
+        String hashedPass = HashUtil.hashPassword(password, user.getSalt());
+        if (user.getPassword().equals(hashedPass)) {
             return user;
         }
         throw new AuthenticationException("Wrong login or password");
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java]: let's modify this method with isPresent() and throw exception only in one place

File: src/main/java/com/internet/shop/util/HashUtil.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        SecureRandom random = new SecureRandom();
+        byte[] salt = new byte[16];
+        random.nextBytes(salt);
+        return salt;
+    }
+
+    public static String hashPassword(String password, byte[] salt) {
+        StringBuilder hashedPassword = new StringBuilder();
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance("SHA512");
```
Comment (dmytro-hryhoruk) [inline src/main/java/com/internet/shop/util/HashUtil.java]: https://mate-academy.github.io/jv-program-fulltime/02_jdbc_and_servlets/checklist/hw_25.html

File: src/main/java/com/internet/shop/util/HashUtil.java (hunk 1 above)
Comment (dmytro-hryhoruk) [reply src/main/java/com/internet/shop/util/HashUtil.java:29]: agree let's throw an exception here

File: src/main/resources/init_dbscheme (hunk 2 above)
Comment (dmytro-hryhoruk) [reply src/main/resources/init_dbscheme]: agree

File: pom.xml (hunk 3 above)
Comment (SetMyDream) [reply pom.xml:40]: it`s used for logger)

File: src/main/java/com/internet/shop/model/User.java (hunk 4 above)
Comment (SetMyDream) [reply src/main/java/com/internet/shop/model/User.java:23]: wow, thanks.
never faced this way declaration before)

File: pom.xml (hunk 3 above)
Comment (dmytro-hryhoruk) [reply pom.xml:40]: yes, but you didn't add any configuration for logging does it work?

Comment (SetMyDream) [reply pom.xml:40]: you right.
Now i added and it works

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**